tests/integrationv2/processes.py

```diff
@@ -170,10 +174,16 @@ def _communicate(self, input_data=None, send_marker_list=None, close_marker=None
                                 input_data_offset = 0
                                 if send_marker_list:
                                     send_marker = send_marker_list.pop(0)
+                                print(f'{self.name}: next send_marker is {send_marker}')
                     elif key.fileobj in (self.proc.stdout, self.proc.stderr):
+                        print(f'{self.name}: stdout available')
                         data = os.read(key.fd, 32768)
                         if not data:
                             selector.unregister(key.fileobj)
```

Comment:
If data is None, does anything else in this iteration need to happen? Like could we do this maybe?
```
if not data:
    selector.unregister(key.fileobj)
    continue
```
Just concerned that there will be "looking for send_marker in None" messages that might be confusing. But maybe having data be None here is not really a case we have to worry about.

Reply:
But we will actually be looking for send_marker in None :) And if it matches, that would totally lead to us sending something.

Bailing early seems like the right behavior to me, but my goal with this PR is not to change the behavior of this method at all. I'm just adding debugging information.

Trying to clean up this method would be a whole other task :) There are many things I think that it's doing wrong, but fixing some of those things would actually break integration tests that rely on the strange behavior.

Reply:
Makes sense!